### PR TITLE
Added periodic adv feature updates in ble 5.3

### PR DESCRIPTION
**Pull request by RoshanESP**
1. Added periodic adv enhancements(adi support) in Bluetooth spec 5.3
2. Current APIs don't have a way to add the parameters for periodic adv commands. So, modified the APIs wherever necessary.

### CODE REVIEW

**Comment by RoshanESP** [comment]
@sjanc  can you PTAL?

File: nimble/host/include/host/ble_gap.h [hunk 1]
```diff
@@ -1524,6 +1524,30 @@ struct ble_gap_periodic_adv_params {
...
+    /** If include adi in aux_sync_ind PDU */
+    unsigned int include_adi:1;
+#endif
+};
+
+/** @brief Periodic advertising sync reporting parameters  */
+struct ble_gap_periodic_adv_sync_report_params {
+#if MYNEWT_VAL(BLE_PERIODIC_ADV_ENH)
+    /** If filter duplicates */
+    unsigned int filter_duplicates:1;
```
Comment (sjanc) [inline nimble/host/include/host/ble_gap.h:1539]: ditto

File: nimble/host/include/host/ble_gap.h [hunk 2]
```diff
@@ -1625,10 +1664,14 @@ int ble_gap_periodic_adv_sync_terminate(uint16_t sync_handle);
...
  * @param enable             If reports should be enabled.
+ * @param params              Additional arguments specifying the particulars
+ *                            of periodic reports.
  *
  * @return                   0 on success; nonzero on failure.
  */
-int ble_gap_periodic_adv_sync_reporting(uint16_t sync_handle, bool enable);
+int ble_gap_periodic_adv_sync_reporting(uint16_t sync_handle,
+                                        bool enable,
+                                        const struct ble_gap_periodic_adv_sync_report_params *params);
```
Comment (sjanc) [inline nimble/host/include/host/ble_gap.h:1674]: same, lest keep params names consistent with function name  ie. ble_gap_periodic_adv_sync_reporting_params

File: nimble/host/src/ble_gap.c [hunk 3]
```diff
@@ -41,6 +41,10 @@
 #define bssnz_t
 #endif
 
+#if MYNEWT_VAL(BLE_PERIODIC_ADV_ENH)
+#define SET_BIT(t, n)  (t |= 1UL << (n))
```
Comment (sjanc) [inline nimble/host/src/ble_gap.c:45]: no need to put macros under #ifdefs

File: nimble/host/src/ble_gap.c [hunk 4]
```diff
@@ -3703,6 +3713,25 @@ ble_gap_periodic_adv_start(uint8_t instance)
...
 }
 
+#if MYNEWT_VAL(BLE_PERIODIC_ADV_ENH)
+static int
+ble_gap_periodic_adv_update_did(uint8_t instance)
+{
+    static uint8_t buf[sizeof(struct ble_hci_le_set_periodic_adv_data_cp)];
+    struct ble_hci_le_set_periodic_adv_data_cp *cmd = (void *) buf;
+    uint16_t opcode;
+    memset(buf, 0, sizeof(buf));
```
Comment (sjanc) [inline nimble/host/src/ble_gap.c:3723]: empty line between variables declarations and code

File: nimble/host/src/ble_gap.c [hunk 5]
```diff
@@ -4215,6 +4281,44 @@ periodic_adv_transfer_enable(uint16_t conn_handle,
...
+    if (!rc) {
+        BLE_HS_DBG_ASSERT(le16toh(rsp.conn_handle) == conn_handle);
+    }
+
+    return rc;
+}
+
+/* BLE_HCI_OCF_LE_SET_DEFAULT_SYNC_TRANSFER_PARAMS command api */
+int
+periodic_adv_set_default_sync_params(uint16_t conn_handle,
```
Comment (sjanc) [inline nimble/host/src/ble_gap.c:4303]: this doesn't look correct,  default sync params shouldn't need connection handle, right?

also ble_hci_le_periodic_adv_sync_transfer_params_cp is not matching BLE_HCI_OCF_LE_SET_DEFAULT_SYNC_TRANSFER_PARAMS opcode...

File: nimble/host/src/ble_gap.c [hunk 6]
```diff
@@ -4215,6 +4281,44 @@ periodic_adv_transfer_enable(uint16_t conn_handle,
...
+    struct ble_hci_le_periodic_adv_sync_transfer_params_cp cmd;
+    struct ble_hci_le_periodic_adv_sync_transfer_params_rp rsp;
+    uint16_t opcode;
+    int rc;
+
+    opcode = BLE_HCI_OP(BLE_HCI_OGF_LE, BLE_HCI_OCF_LE_SET_DEFAULT_SYNC_TRANSFER_PARAMS);
+
+    cmd.conn_handle = htole16(conn_handle);
+    cmd.sync_cte_type = 0x00;
+    cmd.mode = params->reports_disabled ? 0x01 : 0x02;
```
Comment (sjanc) [inline nimble/host/src/ble_gap.c:4315]: and how to disable this?  maybe null params could be used for this?

File: nimble/host/src/ble_gap.c [hunk 7]
```diff
@@ -4215,6 +4281,44 @@ periodic_adv_transfer_enable(uint16_t conn_handle,
...
+
+#if MYNEWT_VAL(BLE_PERIODIC_ADV_ENH)
+    if (params->filter_duplicates)
```
Comment (sjanc) [inline nimble/host/src/ble_gap.c:4318]: this should probably be set only if params->reports_disabled is not set (ie when reports are not disabled)

**Comment by sjanc** [comment]
oh, and while I'm fine with breaking public API here (unfortunate but...) with additions of parameter structure, all samples in repo needs to be updated in same commit to avoid CI breaking

**Comment by RoshanESP** [comment]
Hi @sjanc , Made the suggested changes. PTAL

File: nimble/host/include/host/ble_gap.h [hunk 8]
```diff
@@ -1558,10 +1588,15 @@ int ble_gap_periodic_adv_configure(uint8_t instance,
...
  * @param instance            Instance ID
+ * @param params              Additional arguments specifying the particulars
+ *                            of periodic advertising.
  *
  * @return              0 on success, error code on failure.
  */
-int ble_gap_periodic_adv_start(uint8_t instance);
+int
+ble_gap_periodic_adv_start(uint8_t instance,
+                           const struct ble_gap_periodic_adv_enable_params *params);
```
[the remark this answers is not in the document]
Comment (RoshanESP) [reply nimble/host/include/host/ble_gap.h]: fixed.

File: nimble/host/include/host/ble_gap.h [hunk 9]
```diff
@@ -1524,6 +1524,30 @@ struct ble_gap_periodic_adv_params {
     uint16_t itvl_max;
 };
 
+/** @brief Periodic advertising enable parameters  */
+struct ble_gap_periodic_adv_enable_params {
+#if MYNEWT_VAL(BLE_PERIODIC_ADV_ENH)
+    /** If include adi in aux_sync_ind PDU */
+    unsigned int include_adi:1;
```
[the remark this answers is not in the document]
Comment (RoshanESP) [reply nimble/host/include/host/ble_gap.h]: fixed.

File: nimble/host/include/host/ble_gap.h (hunk 1 above)
Comment (RoshanESP) [reply nimble/host/include/host/ble_gap.h]: Done.

File: nimble/host/include/host/ble_gap.h [hunk 10]
```diff
@@ -1524,6 +1524,30 @@ struct ble_gap_periodic_adv_params {
...
+    /** If filter duplicates */
+    unsigned int filter_duplicates:1;
+#endif
+};
+
+/** @brief Periodic adv set data parameters  */
+struct ble_gap_periodic_adv_set_data_params {
+#if MYNEWT_VAL(BLE_PERIODIC_ADV_ENH)
+    /** If include adi in aux_sync_ind PDU */
+    unsigned int update_did:1;
```
[the remark this answers is not in the document]
Comment (RoshanESP) [reply nimble/host/include/host/ble_gap.h]: Done.

File: nimble/host/include/host/ble_gap.h (hunk 2 above)
Comment (RoshanESP) [reply nimble/host/include/host/ble_gap.h]: Done.

File: nimble/host/include/host/ble_gap.h [hunk 11]
```diff
@@ -1578,10 +1613,14 @@ int ble_gap_periodic_adv_stop(uint8_t instance);
...
  * @param instance            Instance ID
  * @param data                Chain containing the periodic advertising data.
+ * @param params              Additional arguments specifying the particulars
+                             of periodic advertising data.
  *
  * @return          0 on success or error code on failure.
  */
-int ble_gap_periodic_adv_set_data(uint8_t instance, struct os_mbuf *data);
+int ble_gap_periodic_adv_set_data(uint8_t instance,
+                              struct os_mbuf *data,
```
[the remark this answers is not in the document]
Comment (RoshanESP) [reply nimble/host/include/host/ble_gap.h]: Done.

File: nimble/host/src/ble_gap.c (hunk 3 above)
Comment (RoshanESP) [reply nimble/host/src/ble_gap.c:44]: Done.

File: nimble/host/src/ble_gap.c (hunk 4 above)
Comment (RoshanESP) [reply nimble/host/src/ble_gap.c:3722]: Done.

File: nimble/host/include/host/ble_gap.h [hunk 12]
```diff
@@ -1767,10 +1802,14 @@ int ble_gap_periodic_adv_stop(uint8_t instance);
...
  * @param data                Chain containing the periodic advertising data.
+ * @param params              Additional arguments specifying the particulars
+                             of periodic advertising data.
  *
  * @return          0 on success or error code on failure.
  */
-int ble_gap_periodic_adv_set_data(uint8_t instance, struct os_mbuf *data);
+int ble_gap_periodic_adv_set_data(uint8_t instance,
+                                  struct os_mbuf *data,
+                                  struct ble_gap_periodic_adv_set_data_params *params);
```
Comment (sjanc) [inline nimble/host/include/host/ble_gap.h:1812]: could be const

File: nimble/host/src/ble_gap.c [hunk 13]
```diff
@@ -3817,7 +3846,9 @@ ble_gap_periodic_adv_set_data_validate(uint8_t instance,
 }
 
 int
-ble_gap_periodic_adv_set_data(uint8_t instance, struct os_mbuf *data)
+ble_gap_periodic_adv_set_data(uint8_t instance,
+                              struct os_mbuf *data,
+                              struct ble_gap_periodic_adv_set_data_params *params)
```
Comment (sjanc) [inline nimble/host/src/ble_gap.c:3851]: const

**Review by sjanc — COMMENTED** [review]
bttester application also requires update (it was missing in CI check due to misconfiguration of CI target, seehttps://github.com/apache/mynewt-nimble/pull/1629/ )

and some params can be const, otherwise looks OK

File: nimble/host/include/host/ble_gap.h (hunk 12 above)
Comment (RoshanESP) [reply nimble/host/include/host/ble_gap.h]: Done.

File: nimble/host/src/ble_gap.c (hunk 13 above)
Comment (RoshanESP) [reply nimble/host/src/ble_gap.c]: Done.

**Review by sjanc — APPROVED** [review]

**Pull request merged**